### ISSUE TITLE
On Tails (2.12), the electrum-ltc GUI reacts VERY slowly after network connection, but NOT IF the file blockchain_headers is removed before launching

### ISSUE THREAD

**Issue by JazzTp** [pooler/electrum-ltc#48]
Hi,

I'm trying Electrum-LTC from electrum-ltc.org, on Tails 2.12, where I normally use with no problems the latest version (or one of the latest versions) of Electrum for Bitcoin (I do not use the Electrum version coming with Tails which is usually a previous one).

The .electrum-ltc folder is on a pendrive, I launch my script which puts a symlink to it from the home dir, extracts the Electrum-LTC tgz file in the home dir, cds into the folder and launches it from within, ./electrum-ltc or ./electrum (of course I do NOT merge Electrum LTC and BTC!).

Electrum for BTC works perfectly fine this way.

With Electrum-LTC, I have the following issue, both with beta 2.8.3.2 and 2.6.4.2:

- The GUI is responsive if Electrum-LTC is not connected to the network (I configured for automatic connection but can launch it with Tails networking turned off);<br>in this situation, the **top** command shows python2 using 0.3 to 0.7 % CPU
- With Electrum-LTC connected (through Tor just as I do with Electrum for BTC), **the GUI responds very slowly**, various seconds can pass after clicking on a menu item before the pulldown list starts to appear;<br>in this situation, the **top** command shows python2 using 99.9 to 112.2 % CPU, while **other programs remain responsive;**<br>**ctrl-q** closes the GUI (with a delay of several seconds) but the cursor seems not to come back in the terminal window (I waited about 1 min at most, not more) unless I launch from another terminal window a **kill -9 \<process number\>** command.

Here's an example of **terminal window output** with 2.8.3.2; **the first line is the only one which appears if I don't do anything but launching it**, the rest appeared in this specific session when I opened the Network preferences dialog and simply closed it by confirming the current setup):

`Warning: ltc_scrypt not available, using fallback
Traceback (most recent call last):
  File "/home/amnesia/Electrum-LTC-2.8.3.2/gui/qt/main_window.py", line 2726, in run_network_dialog
    NetworkDialog(self.wallet.network, self.config, self).do_exec()
  File "/home/amnesia/Electrum-LTC-2.8.3.2/gui/qt/network_dialog.py", line 51, in do_exec
    self.nlayout.accept()
  File "/home/amnesia/Electrum-LTC-2.8.3.2/gui/qt/network_dialog.py", line 330, in accept
    self.network.set_parameters(host, port, protocol, proxy, auto_connect)
  File "/home/amnesia/Electrum-LTC-2.8.3.2/lib/network.py", line 471, in set_parameters
    self.switch_lagging_interface()
  File "/home/amnesia/Electrum-LTC-2.8.3.2/lib/network.py", line 488, in switch_lagging_interface
    header = self.get_header(self.get_local_height())
  File "/home/amnesia/Electrum-LTC-2.8.3.2/lib/network.py", line 885, in get_header
    return self.blockchain.read_header(tx_height)
  File "/home/amnesia/Electrum-LTC-2.8.3.2/lib/blockchain.py", line 193, in read_header
    f.seek(block_height * 80)
IOError: [Errno 22] Invalid argument
Traceback (most recent call last):
  File "/home/amnesia/Electrum-LTC-2.8.3.2/gui/qt/main_window.py", line 2726, in run_network_dialog
    NetworkDialog(self.wallet.network, self.config, self).do_exec()
  File "/home/amnesia/Electrum-LTC-2.8.3.2/gui/qt/network_dialog.py", line 51, in do_exec
    self.nlayout.accept()
  File "/home/amnesia/Electrum-LTC-2.8.3.2/gui/qt/network_dialog.py", line 330, in accept
    self.network.set_parameters(host, port, protocol, proxy, auto_connect)
  File "/home/amnesia/Electrum-LTC-2.8.3.2/lib/network.py", line 471, in set_parameters
    self.switch_lagging_interface()
  File "/home/amnesia/Electrum-LTC-2.8.3.2/lib/network.py", line 488, in switch_lagging_interface
    header = self.get_header(self.get_local_height())
  File "/home/amnesia/Electrum-LTC-2.8.3.2/lib/network.py", line 885, in get_header
    return self.blockchain.read_header(tx_height)
  File "/home/amnesia/Electrum-LTC-2.8.3.2/lib/blockchain.py", line 193, in read_header
    f.seek(block_height * 80)
IOError: [Errno 22] Invalid argument

`

**Comment by pooler** [comment]
> Warning: ltc_scrypt not available, using fallback

This is the reason for the extreme slowness. Block header verification is much slower for Litecoin than it is for Bitcoin because the hashing algorithm is more complex by almost 3 order of magnitudes. Try installing the `ltc_scrypt` package via pip.

**Comment by JazzTp** [comment]
I see, thank you for replying.
That warning is so important that it could be more verbose and explain what you did to me in your reply, and could also be issued from the GUI.

[Apparently](https://tails.boum.org/contribute/design/persistence/#index6h2), installing pip on Tails might oblige to install Tails on pendrive with Persistence active, which I do for the cold storage part, but I'd like to keep booting the OS supporting the online watching-only part from DVD, and I'd like to be able to just burn a new Tails DVD when new versions are released, and just plug in the pendrive which completely contains the watching-only part of the wallet, as I usually do with Electrum for BTC.

Besides, getting some fundamental part of a wallet from a secondary source is kind of... alarming to me (although it would not touch the cold storage part of the wallet).

**Are there any chances that Electrum-LTC gets to include ltc_scrypt?
That would be just great.**

Besides, let's consider that some persons (personal friends, users who posted Electrum BTC issues) are already protesting or quitting facing the (not so complicated) usage of Electrum... Any further little piece of familiarity with informatics that you add does not help the wallet and the coin in getting adopted, I'm afraid.

(I _do_ realize that using pip on a not-live and not-amnesic Linux installation would be much easier, but being able to run the wallet on Tails has its advantages.)

**Comment by pooler** [comment]
I should point out that package `ltc_scrypt` has C code that needs to be compiled (that's why it speeds things up). For this reason it can't simply be included in Electrum-LTC.

**Comment by JazzTp** [comment]
I see.

Of course I don't know the details of the Electrum-LTC package, but that
seems to me one more (security) reason to include the needed executable
parts, compiled in a secure machine... I think there are other applications
which are distributed in hybrid form, Python + compiled binary... Blender
for instance I believe.

**Comment by JazzTp** [comment]
I see.

Of course I don't know the details of the Electrum-LTC package, but that
seems to me one more (security) reason to include the needed executable
parts, compiled in a secure machine... I think there are other applications
which are distributed in hybrid form, Python + compiled binary... Blender
for instance I believe.

**EDIT:**

Anyway, as the "fallback" word says, I guess the fallback for the absent ltc_scrypt complies and the wallet should work and its GUI get back to responsive after waiting for a while for it to sync with servers, is that correct?

---

On the other hand, I wanted to give it a try to the watching only part in a not live Linux Ubuntu 16.04 LTS... just to be sure I checked that there are no ready made packages for ltc_scrypt or for electrum-ltc (I'd prefer you website as the source anyway), let's try pip to install ltc_scrypt:

```nico@nicopc:~$ sudo pip install ltc_scrypt
[sudo] password for nico: 
The directory '/home/nico/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/home/nico/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
Collecting ltc_scrypt
Exception:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/pip/basecommand.py", line 209, in main
    status = self.run(options, args)
  File "/usr/lib/python2.7/dist-packages/pip/commands/install.py", line 317, in run
    requirement_set.prepare_files(finder)
  File "/usr/lib/python2.7/dist-packages/pip/req/req_set.py", line 360, in prepare_files
    ignore_dependencies=self.ignore_dependencies))
  File "/usr/lib/python2.7/dist-packages/pip/req/req_set.py", line 512, in _prepare_file
    finder, self.upgrade, require_hashes)
  File "/usr/lib/python2.7/dist-packages/pip/req/req_install.py", line 273, in populate_link
    self.link = finder.find_requirement(self, upgrade)
  File "/usr/lib/python2.7/dist-packages/pip/index.py", line 442, in find_requirement
    all_candidates = self.find_all_candidates(req.name)
  File "/usr/lib/python2.7/dist-packages/pip/index.py", line 400, in find_all_candidates
    for page in self._get_pages(url_locations, project_name):
  File "/usr/lib/python2.7/dist-packages/pip/index.py", line 545, in _get_pages
    page = self._get_page(location)
  File "/usr/lib/python2.7/dist-packages/pip/index.py", line 648, in _get_page
    return HTMLPage.get_page(link, session=self.session)
  File "/usr/lib/python2.7/dist-packages/pip/index.py", line 757, in get_page
    "Cache-Control": "max-age=600",
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/sessions.py", line 480, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python2.7/dist-packages/pip/download.py", line 378, in request
    return super(PipSession, self).request(method, url, *args, **kwargs)
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/sessions.py", line 468, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/share/python-wheels/CacheControl-0.11.5-py2.py3-none-any.whl/cachecontrol/adapter.py", line 46, in send
    resp = super(CacheControlAdapter, self).send(request, **kw)
  File "/usr/share/python-wheels/requests-2.9.1-py2.py3-none-any.whl/requests/adapters.py", line 376, in send
    timeout=timeout
  File "/usr/share/python-wheels/urllib3-1.13.1-py2.py3-none-any.whl/urllib3/connectionpool.py", line 610, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/share/python-wheels/urllib3-1.13.1-py2.py3-none-any.whl/urllib3/util/retry.py", line 228, in increment
    total -= 1
TypeError: unsupported operand type(s) for -=: 'Retry' and 'int'
You are using pip version 8.1.1, however version 9.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
nico@nicopc:~$
```

This Ubuntu install is up to date (I ran the software updater minutes before trying this). I also checked that this pip package version is what the synaptic software installer proposes as the up to date package for this Ubuntu desktop distribution.
I would not like to mess it up... but let's give it a try crossing fingers:

```nico@nicopc:~$ sudo pip install --upgrade
The directory '/home/nico/.cache/pip/http' or its parent directory is not owned by the current user and the cache has been disabled. Please check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
The directory '/home/nico/.cache/pip' or its parent directory is not owned by the current user and caching wheels has been disabled. check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
You must give at least one requirement to install (see "pip help install")
You are using pip version 8.1.1, however version 9.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
nico@nicopc:~$
```

**Comment by pooler** [comment]
> Anyway, as the "fallback" word says, I guess the fallback for the absent ltc_scrypt complies and the wallet should work and its GUI get back to responsive after waiting for a while for it to sync with servers, is that correct?

Yes, the slowdown is only due to block header verification during synchronization. The fallback code is pure Python and is therefore about 100 times slower than its C equivalent.

**Comment by JazzTp** [comment]
Thank you.

Tried Electrum-LTC 2.8.3.2 BETA in not-live Ubuntu with ltc_scrypt, ok, it was tending to hang but each time for less than one second.

And... **good news:** using now Electrum-LTC 2.8.3.2 BETA in Tails without ltc_scrypt, it was hanging so I killed the process and **after removing the file ~/.electrum-ltc/blockchain_headers it started to work perfectly fluid, super responsive**, blue circle confirming I'm through Tor (this was OK in Ubuntu too after installing Tor), and I'm now trying a withdrawal from an exchange...

**WOOOOW** the time to write these lines and it's being confirmed already, and the GUI stays perfectly responsive...

First and further confirmations coming muuuch faster than with BTC, wow, **that's just great**, and the exchange is charging a veeery small fee for the withdrawal.
Of course fees might change when LTC will be massively adopted and its value will fly to the moon (now I really think it will, I'll buy in LOL), but I'm confident the second layer using Segwit will provide manner to still be using very low fees. **JUST GREAT congratulations for the nice job :D**

Hey the transaction has already been completely confirmed LOL :D before I could finish this post!

**Comment by JazzTp** [comment]
I could come back to my PC now, some hours have passed so I'm adding a new post, not editing the previous one.

**I confirm the good news:** when the GUI is not very responsive it's certainly **NOT** due to the LTC more intensive hashing algorithm, at least not on my PC, it's like if (just an idea) multiple threads were rushing into opening the blockchain_headers file and process the same data in parallel (or some recursion takes place which should not).

I say this because:

- if at startup this 2.8.3.2 beta version of Electrum-LTC finds the blockchain_headers file, then the GUI lags
- if at startup it does _not_ find the blockchain_headers file, than the GUI is resonsive all the time, also while it is downloading the blockchain_headers file, and later, apparently, processing the headers, or processing something else with the server, until the transaction I had previously made ceases to appear as "Not verified" and gets the green check mark at the left.

**BTW:** while downloading/processing the blockchain_headers file, I think the GUI should show the "synchronizing" circular arrows and possibly also something else indicating that some sync process is happening... Here, otherwise, I only see the test transaction previously made with a **"Not Verified"** label and keep wondering. The GUI could display something like **"Downloading blockchain_headers..."** and later **"Processing blockchain_headers..."**, or **"Downloading blockchain headers related to this wallet transactions and confirmations..."** , and later **"Processing downloaded blockchain headers..."**.

**Remarkably,** the GUI is _perfectly_ responsive even _after_ the download of the blockchain_headers file finishes, and before the "Not Verified" label disappears and the green check mark appears, delay which lasted maybe 1-2 minutes before, just after making that only one transaction, but now, 1-2 hours later, seems to be lasting several minutes...

... It finished now while I'm writing about it, maybe it took 5-10 minutes, if I had expected such a long delay I would have watched the clock to report more accurately.

(The previous size of the blockchain_headers file was 97495120, the current size is 97496400, not a great difference... maybe it had to go through more confirmations online, I have no idea... the transaction details dialogue now says 110 confirmations [already!], by the time I'll finish and post this report they will be much more.)

**Let me guess, the delay is due to the lack of the hybrid Python + compiled C ltc_scrypt package (but the GUI, I repeat myself, stays fluid and responsive).**

**Including ltc_scrypt in the package would be great, indeed.** (As I mentioned, I believe the Blender package is Python+compiled C/C++.)

---

I've sent the amount back to the exchange (to go safer, I renamed the blockchain_headers file before launching, to hide it from the application, so I had the GUI fluid and responsive all the time).

In the options, I had activated 'dynamic fees' and 'propose replaceable always'.

I pressed the Max amount button, and with the default fees, with the cursor in the middle, I got **a warning that higher fees were needed** or the network would not propagate the transaction.

**I guess the warning and/or the dynamic fees algorithms must be revisited, they are evidently not coherent, with dynamic fees active the default fees should not trigger this warning, I suppose.**

With Electrum BTC we can issue a command in the application GUI console to modify the amount of fees/kB the fees cursor can reach... it's probably the same here, but it was not necessary, at least not this time: I moved the fees cursor one step to the right, "within 2 blocks", and the same warning appeared, so I moved it to the extreme right and that went out ok.

The "Replaceable" lable disappeared in one minute and while I'm writing these lines further confirmations are coming in **at incredible speed**, even higher than with the withdrawal.

I was writing something here which I moved elsewhere for clarity and... stopped and wrote **HEY ALREADY CONFIRMED with green check on the left, GO LITECOIN GO, MOOOON LOL... WAIT is this due to Segwit? LTC blocks / confirmations comes out faster than BTC but not THIS faster!!!**

**The fees to send back to the exchange** amounted to 0.00149998 LTC, **about 4 cents of USD** at the moment I'm writing, very low anyway compared to today's BTC fees.

---

Communication, screen-to-webcam, between the two machines with the cold storage and the watching-only part was straightforward, reading the QR code of the unsigned transaction and, back to the online machine, reading the QR code of the signed transaction was instantaneous. 

---

Quitting was instantaneous, too, as I've launched after removing the blockchain_headers file, and control immediately went back to the console. After many minutes on-line, the watching only part only issued these two lines on the console (we already know the first one, which is also the only one the cold storage part issued):

```
Warning: ltc_scrypt not available, using fallback
warning: deprecated tx.hash()
```


---

**I've made two (amazingly fast!!!) transactions, so far this LTC wallet seems to work great, I could only spot for now minor things that need being checked out and re-tuned, not flaws in the basics.**

Huge work certainly behind the scenes, **thank you!!!**

**Comment by JazzTp** [comment]
I just tested Beta 2.8.3.2 on **Tails 3.0**, no ltc_scrypt.

I'm reporting below.

I should probably close this issue (and let developers evaluate whether to reopen it or not), and I should probably open another issue: 

## Segmentation fault when opening the QR code reader.

On Tails 3.0, **Tools | Load transaction | From QR code => segmentation fault**. I don't see the QR code scanner window appear, just the Electrum-LTC window disappear leaving the error on the terminal window from which I had launched it.

I tried with two different webcams, nothing changes. **With Electrum BTC, the QR code scanner works fine with both webcams.**

**This was not happening on Tails 2.12** where I was able to spend LTCs from this wallet using screen-to-webcam communication with the cold storage part.

### New warnings

Electrum-LTC is showing some warnings on Tails 3.0 that I wasn't seeing on Tails 2.12 (Electrum BTC too).

Here they are (the segmentation fault error did not happen unless I tried to open the QR code scanner):

```
Warning: ltc_scrypt not available, using fallback
Gtk-Message: Failed to load module "gail"

** (python2:20455): WARNING **: (bridge.c:993):atk_bridge_adaptor_init: runtime check failed: (root)
_IceTransSocketUNIXConnect: Cannot connect to non-local host localhost.localdomain
_IceTransSocketUNIXConnect: Cannot connect to non-local host localhost.localdomain
Qt: Session management error: Could not open network socket
libpng warning: iCCP: known incorrect sRGB profile
Segmentation fault
```

## Now back to the slow GUI issue, with good news

For the **first test**, I removed the blockchain_headers file. After downloading it, the wallet was showing the transactions as verified again **in less then 10 seconds** (I didn't expect it to be so fast, I had just typed the _date._ command in a terminal window to keep track of when the download had finished, when I saw the green check marks on the left appear, typed _date_ again and that was a 11 seconds time lapse). **Ctrl-q => instantly quit**.

I could made the **second test** a couple of hours later: I didn't remove the blockchain_headers file. The wallet couldn't connect, red circle displayed. I had seen "Tor ready". **Ctrl-q => no quit**, I waited a couple of minutes or less, than killed the process from another terminal window.

I deactivated and activated again the Tails networking, waited for "Tor ready" again, checked with a web browser and...

... The **third test** was **good news: there's no need to remove the blockchain_headers file** and download it again, the GUI is responsive all the time, while the wallet shows "synchronizing" and the two circular arrows are displayed and afterwards. The _top_ command in a terminal window shows python2 as CPU intensive 102-118% with this wallet on this rather old CPU (a Quad), but after a couple of minutes the balance and the blue circle appeared instead of the two circular arrows, and python2 was at 0.3-1% CPU usage. **Ctrl-q => instantly quit**.

With the **following tests**, minutes later, without removing the blockchain_headers file, the syncing process lasted a fraction of a second, I could see the blue circle and the wallet balance almost immediately. **Ctrl-q => instantly quit** (except when I tried to scan a QR code and got the segmentation fault).

**Comment by alxglover** [comment]
Hi @JazzTp, I know this is an old post but I wonder if you could maybe shed some light on exactly how you got Electrum LTC running on Tails in the first place? I've opened a new issue here detailing the problems, which essentially come down to the Linux install instructions specified on the Electrum LTC download page not working properly. Here's my issue: https://github.com/pooler/electrum-ltc/issues/102

If you have any time to post a quick reply, that would be helpful to me and a many others besides.
Thanks.

**Comment by JazzTp** [comment]
Hi,


It is an old post indeed. More recent versions didn't hang that much,
still, in absence of the compiled ltc_scrypt, Electrum-LTC uses the python
fallback, 100 times slower, I think I was told. Which is why I ended up
using the watching only part on Ubuntu instead of Tails, after installing
the compiled ltc_scrypt.

But I understand this is not your request.


How: I DO NOT install Electrum on Tails, I launch Electrum from pendrive
(USB stick) and it runs reading/writing to/from the pendrive, so I can
switch anytime among Tails versions, in case of problems.


A simple bash script I made does the following (this was born for BTC but I
prepared the same for LTC and it works, and I'm ready for when Tails will
also "ship" with Electrum-LTC preinstalled, hopefully with the compiled
ltc_scrypt):


Extracts the Electrum-LTC tree from a password protected compressed file
into the amnesic home dir /home/amnesia

If it exists, eliminates (after issuing a warning and waiting for
confirmation) the .electrum-ltc tree in the amnesic home dir (notice the
dot before electrum-ltc).

Creates a .electrum-ltc symlink to a folder in a pendrive, something like
.../dot/.electrum-ltc-2.9 3.1

Changes dir to the Electrum-LTC tree and launches ./electrum-ltc (important
with BTC to launch from inside the tree otherwise I would be launching the
Electrum version coming preinstalled on Tails, the same might happen for
Electrum-LTC some day soon I think, Electrum-LTC preinstalled on Tails, AND
an Android version).


NOTICE that I maintain the possibility to extract-link-launch different
Electrum-LTC versions, each with its own copy of the
.electrum-LTC-versionInfo folder. I copy wallet files, do not let a new
version convert the only one copy of a wallet file.


How: I simply keep copies of my bash script which names contain the version
info, and inside I replace the version info characters with gedit, the text
editor, with ctrl-h, then I chose which script I launch. The simpler the
better.
A symlink shortens the name for the one I'm adopting as my current default
Electrum version. With some issues, it was super easy to check other
versions and point out which ones presented the issue and which ones didn't


NOTICE that I have separate (encrypted) cold storage / watching only
pendrives and only use the cold storage one on an airgapped machine and
nowhere else (search bad USB exploit, which is the reason why it's
important not to have the cold storage and watching only parts exchange
transactions through pendrive, actually I don't plug anymore in the cold
storage machine ANYTHING which will be plugged elsewhere, I have two
webcams... and in order to upgrade anything on the cold storage, I burn a
CD/DVD).


NOTICE: An issue with the QR scan reader since Tails 3 also appeared with
the Electrum version preinstalled on Tails itself, so apparently it's not
dependent on this way of doing things, oh and it happened on Linux Ubuntu
sometimes too, and to other users with other Linux distributions... I'll
have to report again and ask for more detailed instructions on how to
launch gdb because it gives me a complain, on Ubuntu Linux I mean, on Tails
I don't have it and have no idea how to install, besides I like to use
Tails as it is and boot from DVD to go online).


Cheers

**Comment by alxglover** [comment]
Thank you for your detailed reply, sounds like a pretty effective approach. Perhaps a bit too much faffing for my liking to set up, since it would take me a fair while to replicate these steps without a step-by-step guide and even then I suspect I'd run into issues. I hope in future that Tails comes pre-configured with Electrum LTC, but in lieu of that I think I may just cave and buy a hardware wallet instead for the sake of convenience.

**Comment by JazzTp** [comment]
I see. I might share the script once on my PC, but if you are not familiar
with this Linux things you might end up messing up I'm afraid, although
it's nothing fancy.


Others might be interested in hearing about your experience with the
hardware wallet, maybe you might post a few lines on some forum and put
just a link here (here it is for reporting and discussing issues, I'm
afraid they would remind us... well unless you will be having issues with
the hardware wallet, hopefully not).

**Comment by alxglover** [comment]
Thanks @JazzTp, I much appreciate the information and your time. I have some Linux experience, mainly on web-servers, but I don't generally use it day-to-day in terms of system maintenance. I might have a poke around your script if you end up posting it. Hardware wallets are pretty mature and user friendly these days, so should be pretty simple compared to this. Thanks again.

**Comment by JazzTp** [comment]
Hi @alxglover, _very_ sorry for the delay!

I'm going to put my answer to you into thread #102 from which you referenced this one, so we are not off topic.
